accepted/2024/trusted-publishers-oidc-for-nuget-push.md

```diff
@@ -0,0 +1,568 @@
+first branch of auth (API-key) so considering read-only (package download/restore) scenarios is not needed now.
+
+### Entra ID authentication
+
+We could opt to enable Entra ID service principal-based authentication. Then, users could leverage existing Entra ID
+support for workload identity federation
+([docs](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation)) which uses Federated Identity
+Credentials ([docs](https://learn.microsoft.com/en-us/graph/api/resources/federatedidentitycredentials-overview)). This
+is similar to [VS Marketplace](https://marketplace.visualstudio.com/) and how it could provide extension publishing from
+CI/CD workflows. In the end, this approach could support GitHub Actions via
+[`azure/login`](https://github.com/azure/login) or any compute environment you want (such as an Azure VM with a managed
+identity), as long as you can authenticate as an Entra ID service principal. The drawback of this is that NuGet.org
+would have no control over the workload identity tokens which are accepted and would be opening up a lot more
+authentication scenarios than we really need at this point. This would also require us rationalizing how an Entra ID
+service principal relates to a NuGet.org user or organization profile. Today NuGet.org allows Entra ID (formerly AAD,
+Azure Active Directory) user sign in as well as personal Microsoft account (MSA) sign in. Therefore there is already
+some relationship between NuGet.org and Entra ID. In what way this existing integration would be affected by a service
+principal flow is unclear. So this idea would be a lot more work! Maybe it's something we need in the distant future,
+but not now.
+
+## Prior Art
+
+- This feature has been prototyped and demoed to the NuGet team a couple of times, as far back as February of 2022. More
+  recently, a prototype was made using credential providers and one-legged auth:
+  - [joelverhagen/token-login-sample](https://github.com/joelverhagen/token-login-sample) - sample one-legged auth
+    GitHub Actions workflow
+  - [joelverhagen/token-login](https://github.com/joelverhagen/token-login) - sample one-legged auth step (used by the
+    above)
+  - [NuGet/NuGetGallery@jver-oidc](https://github.com/NuGet/NuGetGallery/tree/jver-oidc) - branch on NuGetGallery to
+    supported one-legged auth
+- [Trusted Publishers for All Package
+Repositories](https://repos.openssf.org/trusted-publishers-for-all-package-repositories), by Seth Larson
+- Trusted Publishers has been implemented by:
+  - Python Package Index, a.k.a. PyPI ([blog](https://blog.pypi.org/posts/2023-04-20-introducing-trusted-publishers/),
+    [docs](https://docs.pypi.org/trusted-publishers/))
+  - RubyGems ([blog](https://blog.rubygems.org/2023/12/14/trusted-publishing.html),
+    [docs](https://guides.rubygems.org/trusted-publishing/))
+  - pub.dev ([docs](https://dart.dev/tools/pub/automated-publishing))
+
+## Unresolved Questions
+
```

Comment:
@joelverhagen , perhaps not directly related to this spec, but something to consider: we raise the security bar for package publishing only when we disable API Key publishing for this package. Following this logic, when a package owner opts-in to OIDC publishing, we should disable API Key publishing for affected packages. There can be many approaches here, and we should have a conversation when we talk about roll out strategy here.

Reply:
@haydentherapper (also participates in OpenSSF) mentioned this here:
https://github.com/NuGet/Home/pull/13673#discussion_r1714296911

I think we should add that as a future opportunity and not plan to do it immediately or in very short order after initial release. Perhaps 3 months+ after initial release or when we feel like we have a strong signal (positive or negative) from our users on how adoption is going in their projects.